## Import Data

In [1]:
import pandas as pd
train_data = pd.read_csv('https://raw.githubusercontent.com/Tariq60/LIAR-PLUS/refs/heads/master/dataset/tsv/train2.tsv', sep = "\t")
train_data.head(1)

0   2635.json      false  \
0  1.0  10540.json  half-true   

  Says the Annies List political group supports third-trimester abortions on demand.  \
0  When did the decline of coal start? It started...                                   

                             abortion    dwayne-bohac State representative  \
0  energy,history,job-accomplishments  scott-surovell       State delegate   

      Texas republican  0.0  1.0  0.0.1  0.0.2  0.0.3         a mailer  \
0  Virginia   democrat  0.0  0.0    1.0    1.0    0.0  a floor speech.   

  That's a premise that he fails to back up. Annie's List makes no bones about being comfortable with candidates who oppose further restrictions on late-term abortions. Then again, this year its backing two House candidates who voted for more limits.  
0  Surovell said the decline of coal "started whe...

In [2]:
# Column 1: the ID of the statement ([ID].json).
# Column 2: the label.
# Column 3: the statement.
# Column 4: the subject(s).
# Column 5: the speaker.
# Column 6: the speaker's job title.
# Column 7: the state info.
# Column 8: the party affiliation.
# Columns 9-13: the total credit history count, including the current statement.
# 9: barely true counts.
# 10: false counts.
# 11: half true counts.
# 12: mostly true counts.
# 13: pants on fire counts.
# Column 14: the context (venue / location of the speech or statement).
# Column 15: the extracted justification

In [3]:
first_data = train_data.columns

In [4]:
train_data.loc[train_data.shape[0]] = first_data
train_data.columns =['index','ID of statement', 'label', 'statement', 'subject', 'speaker', "speaker's job title", 'state info',
                     'party affiliation', 'barely true counts', 'false counts', 'half true counts', 'mostly true counts',
                    'pants on fire counts', 'context', 'extracted justification']
train_data = train_data.drop(columns=['index'])

## Data Cleaning

In [6]:
# label -> clean :)
# subject -> maybe change str to list
# speaker -> clean :)
# speaker's job title -> need to fix big and little letters
# State info -> clean :)
# party affiliation -> clean :)
# barely true counts -> not gonna modify
# false counts -> not gonna modify
# half true counts -> not gonna modify
# mostly true counts -> not gonna modify
# pants on fire counts -> not gonna modify
# context -> assume clean :)
# extracted -> str to list

In [7]:
# subject
train_data['subject'] = train_data['subject'].str.split(",")

# speaker's job title
train_data["speaker's job title"] = train_data["speaker's job title"].str.lower()

# extracted
train_data["extracted justification"] = train_data["extracted justification"].str.split(" ")

train_data.head(3)

ID of statement        label  \
0      10540.json    half-true   
1        324.json  mostly-true   
2       1123.json        false   

                                           statement  \
0  When did the decline of coal start? It started...   
1  Hillary Clinton agrees with John McCain "by vo...   
2  Health care reform legislation is likely to ma...   

                                  subject         speaker speaker's job title  \
0  [energy, history, job-accomplishments]  scott-surovell      state delegate   
1                        [foreign-policy]    barack-obama           president   
2                           [health-care]    blog-posting                 NaN   

  state info party affiliation barely true counts false counts  \
0   Virginia          democrat                0.0          0.0   
1   Illinois          democrat               70.0         71.0   
2        NaN              none                7.0         19.0   

  half true counts mostly true counts pants on fire counts          context  \
0              1.0                1.0                  0.0  a floor speech.   
1            160.0              163.0                  9.0           Denver   
2              3.0                5.0                 44.0   a news release   

                             extracted justification  
0  [Surovell, said, the, decline, of, coal, "star...  
1  [Obama, said, he, would, have, voted, against,...  
2  [The, release, may, have, a, point, that, Miku...

## Factuality Factor

* Social Credibility: People are more likely to perceive a source as credible if others perceive the source is credible
* Stance Detection: What is the political or issue stance of the article or text corpus? How does that affect the veracity of the article or text?
* Stance features: indicate the user's opinion about the news such as supporting or denying.
* Title vs Body: Does the title, agree, discuss, is unrelated to, or negate the body of the text?

#### Social Credibility
* Source History: Delve into the past of the post or source to understand its track record
* Enforsement checks: a post or source that has been enorsed or validated by external reputable entities gains credibility
* revision Analysis: check if the content has been revised updated, or retracted in the past

* Basic info:
    * 10243 total rows in df
    * 4346 types of unique context
    * top five context
        * a news release                                                  241
        * an interview                                                    229
        * a press release                                                 223
        * a speech                                                        214
        * a TV ad                                                         180

In [51]:
# all imports
import numpy as np
import scipy
import sklearn
import keras

from sklearn.preprocessing import OneHotEncoder
ohe = OneHotEncoder(handle_unknown = "ignore", sparse = False)

# citation: https://scikit-learn.org/dev/modules/generated/sklearn.preprocessing.OneHotEncoder.html

In [52]:
train_data['party affiliation'].unique()

array(['democrat', 'none', 'republican', 'organization', 'independent',
       'columnist', 'activist', 'talk-show-host', 'libertarian',
       'newsmaker', 'journalist', 'labor-leader', 'state-official',
       'business-leader', 'education-official', 'tea-party-member', nan,
       'green', 'liberal-party-canada', 'government-body', 'Moderate',
       'democratic-farmer-labor', 'ocean-state-tea-party-action',
       'constitution-party'], dtype=object)

In [54]:
# spliting up the training data set to 8:2
small_training =  train_data[0: int(train_data.shape[0]*0.8)]
small_testing =  train_data[int(train_data.shape[0]*0.8):int(train_data.shape[0])]

# dropping null
small_training = small_training.dropna()
small_testing = small_testing.dropna()

# keeping only 4 columns
small_training = small_training[["label", "speaker", "context", "party affiliation"]]
small_testing = small_testing[["label", "speaker", "context", "party affiliation"]]

# ohe (training data)
ohe_label = ohe.fit_transform(np.array(small_training['label']).reshape(-1,1))
small_training_ohe_label_df = pd.DataFrame(ohe_label, columns = list(small_training['label'].unique()))

ohe_speaker = ohe.fit_transform(np.array(small_training['speaker']).reshape(-1,1))
ohe_context = ohe.fit_transform(np.array(small_training['context']).reshape(-1,1))
ohe_party = ohe.fit_transform(np.array(small_training['party affiliation']).reshape(-1,1))
ohe_speaker_ohe_context_ohe_party = []
for i in range(len(ohe_speaker)):
    ohe_speaker_ohe_context_ohe_party.append(np.concatenate((ohe_speaker[i], ohe_context[i], ohe_party[i])))
ohe_speaker_ohe_context_ohe_party = np.array(ohe_speaker_ohe_context_ohe_party)

small_training_ohe_speaker_context_party_df = pd.DataFrame(ohe_speaker_ohe_context_ohe_party, columns = 
                                             list(small_training['speaker'].unique()) 
                                             + list(small_training['context'].unique())
                                                + list(small_training['party affiliation'].unique()))

# ohe (testing data)
ohe_label = ohe.fit_transform(np.array(small_testing['label']).reshape(-1,1))
small_testing_ohe_label_df = pd.DataFrame(ohe_label, columns = list(small_testing['label'].unique()))

ohe_speaker = ohe.fit_transform(np.array(small_testing['speaker']).reshape(-1,1))
ohe_context = ohe.fit_transform(np.array(small_testing['context']).reshape(-1,1))
ohe_party = ohe.fit_transform(np.array(small_testing['party affiliation']).reshape(-1,1))
ohe_speaker_ohe_context_ohe_party = []
for i in range(len(ohe_speaker)):
    ohe_speaker_ohe_context_ohe_party.append(np.concatenate((ohe_speaker[i], ohe_context[i], ohe_party[i])))
ohe_speaker_ohe_context_ohe_party = np.array(ohe_speaker_ohe_context_ohe_party)


small_testing_ohe_speaker_context_party_df = pd.DataFrame(ohe_speaker_ohe_context_ohe_party, columns = 
                                             list(small_testing['speaker'].unique()) 
                                             + list(small_testing['context'].unique())
                                                + list(small_testing['party affiliation'].unique()))
# citation: https://scikit-learn.org/dev/modules/generated/sklearn.preprocessing.OneHotEncoder.html

In [61]:
# algorithm/method #1
from keras.models import Sequential
from keras import layers
from keras.layers import Dense

nn = keras.Sequential([
    keras.Input(shape=(small_training_ohe_speaker_context_party_df.shape[1])),
    layers.Dense(10, activation='sigmoid'),
    layers.Dense(10, activation='sigmoid'),
    layers.Dense(10, activation='sigmoid'),
    layers.Dense(6),
])

nn.compile(optimizer = keras.optimizers.Adam(learning_rate = 0.01), loss = 'mse', metrics = "accuracy")

nn.fit(small_training_ohe_speaker_context_party_df, small_training_ohe_label_df, batch_size = 10, epochs = 100)

# citation: https://keras.io/guides/sequential_model/

Epoch 1/100
534/534 [==============================] - 3s 4ms/step - loss: 0.1414 - accuracy: 0.2116
Epoch 2/100
534/534 [==============================] - 2s 3ms/step - loss: 0.1328 - accuracy: 0.2801
Epoch 3/100
534/534 [==============================] - 2s 3ms/step - loss: 0.1195 - accuracy: 0.4245
Epoch 4/100
534/534 [==============================] - 2s 3ms/step - loss: 0.0943 - accuracy: 0.6123
Epoch 5/100
534/534 [==============================] - 2s 3ms/step - loss: 0.0733 - accuracy: 0.6650
Epoch 6/100
534/534 [==============================] - 2s 3ms/step - loss: 0.0650 - accuracy: 0.6873
Epoch 7/100
534/534 [==============================] - 2s 4ms/step - loss: 0.0607 - accuracy: 0.7211
Epoch 8/100
534/534 [==============================] - 2s 3ms/step - loss: 0.0574 - accuracy: 0.7316
Epoch 9/100
534/534 [==============================] - 2s 3ms/step - loss: 0.0541 - accuracy: 0.7490
Epoch 10/100
534/534 [==============================] - 2s 4ms/step - loss: 0.0527 - accura

### Findings: when use both context and speaker, the accuracy of using a simple neural network with sigmoid as the activation function was higher (~32% to 66%)

In [43]:
# algorithm/method #2
nn = keras.Sequential([
    keras.Input(shape=(small_training_ohe_speaker_context_df.shape[1])),
    layers.Dense(10, activation='relu'),
    layers.Dense(10, activation='relu'),
    layers.Dense(10, activation='relu'),
    layers.Dense(6),
])

nn.compile(optimizer = keras.optimizers.Adam(learning_rate = 0.01), loss = 'mse', metrics = "accuracy")

nn.fit(small_training_ohe_speaker_context_df, small_training_ohe_label_df, batch_size = 10, epochs = 50)

# citation: https://keras.io/guides/sequential_model/

Epoch 1/50
534/534 [==============================] - 3s 4ms/step - loss: 0.1374 - accuracy: 0.2002
Epoch 2/50
534/534 [==============================] - 2s 4ms/step - loss: 0.1369 - accuracy: 0.2048
Epoch 3/50
534/534 [==============================] - 2s 4ms/step - loss: 0.1369 - accuracy: 0.2058
Epoch 4/50
534/534 [==============================] - 2s 3ms/step - loss: 0.1369 - accuracy: 0.2032
Epoch 5/50
534/534 [==============================] - 2s 4ms/step - loss: 0.1369 - accuracy: 0.2043
Epoch 6/50
534/534 [==============================] - 2s 4ms/step - loss: 0.1368 - accuracy: 0.2088
Epoch 7/50
534/534 [==============================] - 2s 4ms/step - loss: 0.1369 - accuracy: 0.1983
Epoch 8/50
534/534 [==============================] - 2s 4ms/step - loss: 0.1369 - accuracy: 0.2011
Epoch 9/50
534/534 [==============================] - 2s 4ms/step - loss: 0.1370 - accuracy: 0.2050
Epoch 10/50
534/534 [==============================] - 2s 3ms/step - loss: 0.1368 - accuracy: 0.2054

### Findings: when use both context and speaker, the accuracy of using a simple neural network with relu as the activation function was lower (~32% to 20%)

In [45]:
# algorithm/method #3
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
dt = DecisionTreeClassifier(criterion= "entropy", splitter = "best")

accuracy = cross_val_score(dt, small_training_ohe_speaker_context_df, small_training_ohe_label_df, cv=5)

accuracy

# citation: https://scikit-learn.org/dev/modules/generated/sklearn.tree.DecisionTreeClassifier.html
# https://scikit-learn.org/stable/modules/cross_validation.html

array([0.18650422, 0.20075047, 0.19043152, 0.18761726, 0.18949343])

### Findings: when use both context and speaker, the accuracy of using a simple decision tree got a bit higher (~6% to ~19%)

#### Stance Detection (Political Affiliation)
* Language Inspection: Scrutinize content for language indicative or political inclination
* Disclosure Checks: Ensure any affiliations by the author or source are openly disclosed
* Fact-checker comparison: contrast content claims against neutal, non-partisan fact-checkers

#### Stance Features (Political Bias)
* Bias Detection: Determine if content distorts facts to favor a specific political entity
* Tonal Analysis: review tone and language for signs of political partiality
* Event Portrayal: ensure political events are represented without consistent bias.

#### Title vs Body (Dont see anything similar to this on the pdf...)
